### PR TITLE
App Premium implementation

### PR DESCRIPTION
**Pull request by Luc1412**
## Summary

This represent the implementation of [App Subscriptions/App Premium](https://discord.com/developers/docs/monetization/overview). 

I was already able to test the implementation, since I got invited to join the Beta. Based on my tests, everything should work flawlessly. 

This PR may need improvements regarding util methods, which interconnect models and simplify the workflow with entitlements.

## Checklist

- [ ] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/sku.py
Comment (Rapptz) [inline discord/sku.py:83]:
```suggestion
    user_id: Optional[:class:`int`]
```

File: discord/sku.py
Comment (Rapptz) [inline discord/sku.py:92]:
```suggestion
        Whether the entitlement has been consumed. Not applicable to app subscriptions so will typically be ``False``.
```

File: discord/sku.py
Comment (Rapptz) [inline discord/sku.py:100]: These two fields are undocumented. They are in the examples though, so it's probably okay.

File: discord/sku.py
Comment (Rapptz) [inline discord/sku.py:126]:
```suggestion
        self.consumed: bool = data.get('consumed', False)
```

File: discord/sku.py
Comment (Rapptz) [inline discord/sku.py:129]:
```suggestion
        self.deleted: bool = data.get('deleted', False)
```

File: discord/sku.py
Comment (Rapptz) [inline discord/sku.py:72]: This type needs a few things for ergonomics:

1. A `user` property similar to `guild`.
2. Probably an `sku` property that gives either an `SKU` or `PartialSKU` type?
3. An `is_valid()` method to check if the entitlement is still valid
4. The `delete()` method from earlier.

**Review by Rapptz — CHANGES_REQUESTED** [review]
Thanks for the PR. This is a first pass review at this.

File: discord/sku.py
Comment (Luc1412) [reply discord/sku.py:72]: 2.) I can't create an SKU, hence I'm missing crucial information. A PartialSKU would just store the ID. Not sure if it worth it to create such class.
3.) What is valid? Just checking the expiry, then `is_expired()` might be better. A future proof `valid` method could include the consumed attribute, if Discord adds support for one-time purchases or uses the same api for the new server shop feature

File: discord/sku.py
Comment (Luc1412) [reply discord/sku.py:100]: I think they can be removed since they add no value. `deleted` should be always false, since when the sku is deleted, is should just disappear.

File: discord/client.py
Comment (Luc1412) [reply discord/client.py]: Not sure if it's intended but changing List[:class:`.SKU`] to List[:class:`SKU`] causes trouble with the docs

File: discord/sku.py
Comment (Luc1412) [reply discord/sku.py:102]: Talked with Mason. Those attributes won't provide any value

File: discord/sku.py
Comment (Rapptz) [inline discord/sku.py:155]:
```suggestion
        if self.ends_at is None:
            return False
        return utils.utcnow() >= self.ends_at
```

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: discord/enums.py
Comment (Puncher1) [inline discord/enums.py:74]:
```suggestion
    'EntitlementType',
    'EntitlementOwnerType',
```

File: discord/flags.py
Comment (Puncher1) [inline discord/flags.py:1]: https://github.com/Rapptz/discord.py/blob/c86624df4b7a94c9936ed033125f1e03324b3f94/discord/flags.py#L62
Add `'SKUFlags'` to `__all__`

File: discord/flags.py
Comment (Puncher1) [inline discord/flags.py:2032]:
```suggestion

    @flag_value
    def available(self):
        """:class:`bool`: Returns ``True`` if the SKU is available for purchase."""
        return 1 << 2
        
```
See https://github.com/discord/discord-api-docs/pull/6477

File: discord/sku.py
Comment (Puncher1) [inline discord/sku.py:1]: Header comment is missing.

File: discord/sku.py
Comment (Puncher1) [inline discord/sku.py:99]:
```suggestion
    deleted: :class:`bool`
        Whether the entitlement has been deleted.
```
See https://github.com/discord/discord-api-docs/pull/6477

File: discord/sku.py
Comment (Puncher1) [inline discord/sku.py:114]:
```suggestion
        'deleted',
```
See https://github.com/discord/discord-api-docs/pull/6477

File: discord/sku.py
Comment (Puncher1) [inline discord/sku.py:127]:
```suggestion
        self.deleted: bool = data['deleted']
```
See https://github.com/discord/discord-api-docs/pull/6477

File: discord/types/sku.py
Comment (Puncher1) [inline discord/types/sku.py:1]: Header comment is missing.

File: docs/api.rst
Comment (Puncher1) [inline docs/api.rst:4797]: `SKUFlags` is missing.

**Review by Puncher1 — COMMENTED** [review]
* Updated fields from following PR: https://github.com/discord/discord-api-docs/pull/6477
* Improvement suggestions

**Comment by Luc1412** [comment]
> * Updated fields from following PR: [Premium Apps: SKU and Entitlement fields discord/discord-api-docs#6477](https://github.com/discord/discord-api-docs/pull/6477)

I'd say we better wait for an official approval of this PR.

**Comment by Rapptz** [comment]
That PR is by an employee.

**Comment by Luc1412** [comment]
> That PR is by an employee.

My bad. Missed the contributor tag. Then gonna add these changes.

File: discord/state.py
Comment (Rapptz) [inline discord/state.py:1592]:
```suggestion
    def parse_entitlement_update(self, data: gw.EntitlementUpdateEvent) -> None:
```

File: discord/state.py
Comment (Rapptz) [inline discord/state.py:1596]:
```suggestion
    def parse_entitlement_delete(self, data: gw.EntitlementDeleteEvent) -> None:
```

File: docs/api.rst
Comment (Rapptz) [inline docs/api.rst:534]:
```suggestion
    :param entitlement: The entitlement that was deleted.
```

File: docs/api.rst
Comment (Rapptz) [inline docs/api.rst:508]:
```suggestion
    :param entitlement: The entitlement that was created.
```

File: docs/api.rst
Comment (Rapptz) [inline docs/api.rst:525]:
```suggestion
    Called when a users subscription to a SKU is cancelled. This is typically only called when:

    - Discord issues a refund for the subscription.
    - Discord removes an entitlement from a user.
```

File: docs/api.rst
Comment (Rapptz) [inline docs/api.rst:530]:
```suggestion
    .. warning::

        This event won't be called if the user cancels their subscription manually, instead
        :func:`on_entitlement_update` will be called with :attr:`Entitlement.ends_at` set to the end of the
        current billing period.
```

File: docs/api.rst
Comment (Rapptz) [inline docs/api.rst:514]:
```suggestion
    Called when a user updates their subscription to a SKU. This is usually called when
    the user renews or cancels their subscription.
```

**Review by Rapptz — CHANGES_REQUESTED** [review]
Minor doc nits.

There's a bit of a gotcha with the design if someone decides to save Entitlement instances to a variable somewhere since those instances aren't ever updated with gateway information those instances will became stale. It's not an issue with this PR, but it's something to keep in mind.

**Review by Rapptz — APPROVED** [review]
Thanks.

**Pull request merged**
